core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,298 @@
+package com.microsoft.applicationinsights.internal.heartbeat;
+
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.TelemetryConfiguration;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.internal.shutdown.Stoppable;
+import com.microsoft.applicationinsights.telemetry.MetricTelemetry;
+import com.microsoft.applicationinsights.telemetry.Telemetry;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+
+/**
+ * <p>
+ *  Concrete implementation of Heartbeat functionality. This class implements
+ *  {@link com.microsoft.applicationinsights.internal.heartbeat.HeartBeatProviderInterface} and
+ *  {@link com.microsoft.applicationinsights.internal.shutdown.Stoppable}
+ * </p>
+ *
+ * @author Dhaval Doshi
+ * @since 03-30-2018
+ */
+public class HeartBeatProvider implements HeartBeatProviderInterface, Stoppable {
+
+  /**
+   * The name of the heartbeat metric.
+   */
+  private final String HEARTBEAT_SYNTHETIC_METRIC_NAME = "HeartbeatState";
+
+  /**
+   * The list of disabled properties
+   */
+  private List<String> disableDefaultProperties = new ArrayList<>();
+
+  /**
+   * List of disabled heartbeat providers
+   */
+  private List<String> disabledHeartBeatPropertiesProviders = new ArrayList<>();
+
+  /**
+   * The counter for heartbeat sent to portal
+   */
+  private AtomicLong heartbeatsSent;
+
+  /**
+   * Map to hold heartbeat properties
+   */
+  private ConcurrentMap<String, HeartBeatPropertyPayload> heartbeatProperties;
+
+  /**
+   * Interval at which heartbeat would be sent
+   */
+  private long interval;
+
+  /**
+   * Telemetry client instance used to send heartbeat.
+   */
+  private TelemetryClient telemetryClient;
+
+  /**
+   * ThreadPool used for adding properties to concurrent dictionary
+   */
+  private ExecutorService executorService = Executors.newCachedThreadPool();
+
+  /**
+   * Threadpool used to send data heartbeat telemetry
+   */
+  private ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
```

Comment:
If you're going to initialize this with only 1 thread allowed make sure the Runnables allowed in it will always complete before the next one is scheduled to run. This also depends if this pool works on a timer or a delay (it could support both).

What happens if heartbeat@t_0 is stuck in the pool and heartbeat@t_1 gets scheduled?

Reply:
In a way you are correct. This can happen, considering the situation --> Thread A (starts sending heartbeat) --->current-state(blocked in wire) ---> ThreadPool gets instruction to schedule a new heartbeat and it would be in blocking state unless it gets a free thread to perform the task. This is practically unavoidable since there is no ScheduledCacheThreadPool. 

However, performing the math, considering the retry time of 15 minutes and the worst case scenario where heartbeat is scheduled for every 30 minutes if we set the maxPoolSize of 30 it would suffice for our scenario and ensure that every heartbeat gets a worker in any case. After that retries would be over and we would have a free worker thread to pick the new HeartBeat payload. I have pushed this changes, let me know your thoughts here.

Reply:
One more thing I found that ScheduledThreadPool doesn't have the concept of corePool and maxPool size. So in this case we might need to specify 'X' threads by default which will be always available to complete the task.